app/models/champs/multiple_drop_down_list_champ.rb

```diff
@@ -55,12 +57,58 @@ def value=(value)
     end
   end
 
+  def store_referentiels
+    self.referentiels = referentiels_from(value)
+  end
+
+  def referentiel_items_selected?
+    referentiels.present?
+  end
+
+  def referentiels_items_column_values
+    return [] if referentiels.nil?
+    referentiels.map do |referentiel|
+      referentiel_headers.map { |(header, path)| [header, referentiel.second.dig('data', 'row', path)] }
+    end
+  end
+
+  def referentiels_items_first_column_values
+    referentiels_items_column_values.map { |referentiel| referentiel.first.second }
+  end
+
+  def referentiel_headers
+    return [] if referentiels&.blank? || referentiels&.first&.blank?
+
+    headers = referentiels.first.second&.dig('data', 'headers') || []
+    headers.map { |header| [header, Referentiel.header_to_path(header)] }
+  end
+
   private
 
+  def referentiels_from(value)
+    return if value.blank?
+    values = JSON.parse(value)
+    referentiels_data = {}
+    values.each do |id|
+      referentiel_item = type_de_champ.referentiel.items.find_by(id: id.to_i)
+
+      # When changing tdc type or simple/advanced mode, champ value is not an item id
+      if referentiel_item.blank?
+        self.value = nil
+        return
+      end
+
+      headers = referentiel_item.referentiel.headers
+      referentiels_data = referentiels_data.merge(id => { data: referentiel_item.data.merge(headers:) })
+    end
+    referentiels_data
+  end
```

Comment:
2/3 trucs me chagrinait dans cette methode : 
1. json.parse sans rescue, je suis pas sur que ca nous pete pas a la gueule un jour. Est ce qu'on souhaite bien retourner un {} ?
2. find_by dans une itereation. 
3. l'iterateur avec un objet intermediaire. each_with_object rend la syntax plus lisible a mon avis

```suggestion
  def referentiels_from(value)
    return if value.blank?
    values = JSON.parse(value)
    referentiel_items = type_de_champ.referentiel.items.where(id: values)
    if referentiel_items.empty?
      self.value = nil
    else
      referentiel_items.each_with_object({}) do |item, referentiels_data|
        headers = item.referentiel.headers
        referentiels_data[item.id] = { data: item.data.merge(headers:) }
      end
    end
  rescue JSON::ParserError
    {}
  end
```

Reply:
oui, ça m'a l'air bien, je vais tester